### PR TITLE
[20608][20610] Address XMLProfiles fuzzer regressions

### PR DESCRIPTION
**Pull request by Mario-DL**
## Description

This PR addresses the regressions described in [20608](https://eprosima.easyredmine.com/issues/20608) and [20610](https://eprosima.easyredmine.com/issues/20610). One has to  do with catching an out of bounds exception and the other with fixing the regex pattern.

## Contributor Checklist
- [X] Commit messages follow the project guidelines. 
- [X] The code follows the style guidelines of this project. 
- [X] Tests that thoroughly check the new feature have been added/Regression tests checking the bug and its fix have been added; the added tests pass locally 
- **N/A** Any new/modified methods have been properly documented using Doxygen. 
- [X] Changes are ABI compatible. 
- [X] Changes are API compatible. 
- **N/A** New feature has been added to the `versions.md` file (if applicable).
- **N/A** New feature has been documented/Current behavior is correctly described in the documentation. 
    
- **N/A** Applicable backports have been included in the description.

## Reviewer Checklist
- [x] The PR has a milestone assigned.
- [x] The title and description correctly express the PR's purpose.
- [x] Check contributor checklist is correct.
- [x] Check CI results: changes do not issue any warning.
- [x] Check CI results: failing tests are unrelated with the changes.

### CODE REVIEW

File: src/cpp/utils/UnitsParser.cpp [hunk 1]
```diff
@@ -77,7 +77,7 @@ uint32_t parse_value_and_units(
 
     to_uppercase(units);
 
-    std::regex pattern(R"(([kibmgKIBMG]{1,3})|(\d+))");
+    std::regex pattern(R"(([bB]{1})|([kibmgKIBMG]{2,3})|(\d+))");
```
Comment (MiguelCompany) [inline src/cpp/utils/UnitsParser.cpp:80]: 1. We already converted to uppercase, so we don't need the lowercase letters.
2. The 'at()` below is doing a search in the map (i.e. performing string comparisons) so, in a way, it is doing the job of the regex again.

This means we can remove the regex altogether, and just capture `out_of_range` around the `at()` operation, throwing the `invalid_argument` we have in line 83

An additional update we could make is adding the empty string to the map, with the same value we have for `"B"`, and remove the conditionals in the places where `parse_value_and_units` is being called

**Review by MiguelCompany — CHANGES_REQUESTED** [review]

File: src/cpp/utils/UnitsParser.cpp (hunk 1 above)
Comment (Mario-DL) [reply src/cpp/utils/UnitsParser.cpp]: I like this suggestion

File: src/cpp/utils/UnitsParser.cpp [hunk 2]
```diff
@@ -64,16 +65,29 @@ uint32_t parse_value_and_units(
...
+    {
+        throw std::invalid_argument("Failed to parse value from string." \
+                      " The number is too large to be converted to bytes (Max: (2^32)-1 Bytes).");
+    }
+
     to_uppercase(units);
 
-    std::regex pattern(R"(([kibmgKIBMG]{1,3})|(\d+))");
-    if (!std::regex_match(units, pattern))
+    unsigned int magnitude = 0;
```
Comment (MiguelCompany) [inline src/cpp/utils/UnitsParser.cpp:81]:
```suggestion
    uint32_t magnitude = 0;
```

**Review by MiguelCompany — CHANGES_REQUESTED** [review]

**Review by MiguelCompany — APPROVED** [review]

**Comment by Mario-DL** [comment]
@richiprosima please test this

**Pull request merged**